### ISSUE TITLE
Token-level Policy Gradient Loss

### ISSUE THREAD

**Issue by gongel** [BytedTsinghua-SIA/DAPO#4]
你好，有些疑问，麻烦解答下。
verl 里面 Policy Loss 的实现默认就是Token-level Policy Gradient Loss 吧？

**Comment by jszyxw00** [comment]
code中use_token_level_loss的含义与对应逻辑是不是反了？

    if use_token_level_loss:
        pg_loss = verl_F.masked_mean(pg_losses, eos_mask)
    else:
        pg_loss = torch.sum(pg_losses * eos_mask, dim=1) / seq_len_per_sample
        pg_loss = torch.mean(pg_loss)

**Comment by gongel** [comment]
应该没反，else 的情况就是 sample-level 了，if 就是目前 verl 里面默认的 token-level。

**Comment by huiyeruzhou** [comment]
> 应该没反，else 的情况就是 sample-level 了，if 就是目前 verl 里面默认的 token-level。

That's right!

**Comment by yqy2001** [comment]
> 你好，有些疑问，麻烦解答下。 verl 里面 Policy Loss 的实现默认就是Token-level Policy Gradient Loss 吧？

Yes. But many previous RL algorithms use sample-level loss that takes expectation over trajectories.
